src/mmda/predictors/heuristic_predictors/dictionary_word_predictor.py

```diff
@@ -80,113 +124,301 @@ def predict(self, document: Document) -> List[SpanGroup]:
+            max_row_end_token_id_to_min_row_start_token_id=max_row_end_token_id_to_min_row_start_token_id,
+            punct_strip_candidate_token_ids=punct_strip_candidate_token_ids
+        )
+
+        # 4) transformation
+        words: List[SpanGroup] = self._convert_to_words(
+            document=document,
+            token_id_to_word_id=token_id_to_word_id,
+            word_id_to_text=word_id_to_text
+        )
+
+        # 5) cleanup
+        document.remove(field_name='_ws_tokens')
+        return words
 
-                words.append(self._copy_token_with_text(token))
+    def _precompute_token_features(self, document: Document) -> Tuple:
+        """
+        Compute stuff necessary for dictionary-building and/or merging tokens into words.
 
-            if len(curr_row.tokens) == 0:
-                raise ValueError("Encountered row without any tokens!")
+        1. `whitespace_tokenization` is necessary because lack of whitespace is an indicator that
+           adjacent tokens belong in a word together.
 
-            curr_row_last_token = curr_row.tokens[-1]
+        2. `beginning|end_of_row|page` is necessary because row transitions are often where tokens
+           should be merged into a word. Knowing this also helps with determining what are "safe"
+           words to add to dictionary.
 
-            # If we are at the end of the document just add token
-            if not next_row:
-                words.append(self._copy_token_with_text(curr_row_last_token))
+        3. `punctuation` in `start_of_row` tokens is necessary because we may need to keep them
+        as separate tokens even if there is a word merge (e.g. the semicolon "fine-tuning;")
+        """
+        _ws_tokens: List[SpanGroup] = self.whitespace_predictor.predict(document=document)
+        document.annotate(_ws_tokens=_ws_tokens)
+
+        # token -> ws_tokens
+        token_id_to_ws_token_id = {}
+        for token in document.tokens:
+            token_id_to_ws_token_id[token.id] = token._ws_tokens[0].id
+
+        # ws_token -> tokens
+        ws_token_id_to_tokens = defaultdict(list)
+        for token_id, ws_token_id in token_id_to_ws_token_id.items():
+            ws_token_id_to_tokens[ws_token_id].append(token_id)
+
+        # token -> all cluster tokens
+        token_id_to_token_ids = {}
+        for token_id, ws_token_id in token_id_to_ws_token_id.items():
+            candidate_token_ids = [i for i in ws_token_id_to_tokens[ws_token_id]]
+            token_id_to_token_ids[token_id] = candidate_token_ids
+
+        # beginning/end of row w/ hyphen
+        # TODO: add pages too
+        row_end_with_hyphen_token_ids = set()
+        row_start_after_hyphen_token_ids = set()
+        max_row_end_token_id_to_min_row_start_token_id = {}
+        for i in range(0, len(document.tokens) - 1):
+            current = document.tokens[i]
+            next = document.tokens[i + 1]
+            is_transition = current.rows[0].id != next.rows[0].id
```

Comment:
do we have to update pdfplumber for this? does it not work if we say like `current.rows[0] != next.rows[0]`? just so we don't have to update pdfplumber as well

Reply:
yea, unfortunately this would make things much easier to write